mautic/mautic
# Segment membership as a new filter for dynamic web content

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | #2856
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Add ability to use segment filters in dynamic content.

#### Steps to test this PR:
0. Not needed in beta - Run ` dcp bin/console mautic:assets:generate` if you don't want to use dev local env.  
1. Create segment to match contact in step 5
2. Create dynamic content
    - `Is campaign based?` = NO
![Snímek obrazovky 2020-10-29 v 12 31 39](https://user-images.githubusercontent.com/12815758/97563152-f04fd880-19e2-11eb-8dfc-000df2c14d27.png)
    - then click filter in the top > select `Segment membership` 
    - use existing segment in which is created contact added
![Snímek obrazovky 2020-10-29 v 12 31 50](https://user-images.githubusercontent.com/12815758/97563166-f3e35f80-19e2-11eb-9132-6ea4f1f0a5a7.png)
3. Create dynamic content page in `Components > Landing pages`
    - Fill required info
    - Go to `Builder`
    - Drag `Dynamic content` into page content
    - Click on it
    - In right box select your dynamic content by name and use `test` as requested slot name
![Snímek obrazovky 2020-10-29 v 12 36 43](https://user-images.githubusercontent.com/12815758/97563571-85eb6800-19e3-11eb-988b-782caf918172.png)
    - Close builder and save
    - with url `https://yourmautic.coml/test` check landing page
    - You 'll see content from step 5
4. Create form with possibility to add contact `Components > Forms` 
    - mapped and required fields Firstname, Lastname, Email
5. Use this form in anonymous browser session 
    - with url `https://mautic-cloud.local/form/[ID]` to add contact with criteria matching created segment in step 1
    - keep this browser window open
    - open new tab with landing page from step 3
    - You'll see dynamic content defined in step 2

#### Other areas of Mautic that may be affected by the change:
1. `CustomObjectsBundle\EventListener\DynamicContentSubscriber::evaluateFilters()` is using `stopPropagation` method. This could eliminate further listeners from running if incorrect priority is set- https://github.com/mautic-inc/plugin-custom-objects/blob/b2ed6c064e3e256f2640fdb456bbf2cfb879300b/EventListener/DynamicContentSubscriber.php#L117

#### TODO
- [x] Frontend (form handling, js)
- [x] `PHP Notice - Array to string conversion``$choice` contains [] in 
```
/vendor/symfony/form/ChoiceList/ArrayChoiceList.php:68 at 
```
When saving form kwith operator `empty`/`not empty`. Selected value before is saved in the same way as in lead_list table
- [x] Landing pages are not working for me with DWC. John has no prob. Confirmed as working in my beta instance. Something wrong in my docker env. - ,Mistake in DC & landing page config
- [x] Implementation into landing pages
- [x] ~~Functional test for landing page~~ removed here https://github.com/mautic-inc/mautic-cloud/pull/1037/commits/86ac47c73891ee6561caccf4589949920b4e514a. HTTP requests from local are blocked in staging.
